### PR TITLE
[WIP] Minimal implementation without numba_scipy

### PR DESCRIPTION
**Pull request by StefanUlbrich**
Use naïve implementation of `gammaln` for positive integers instead of `scipy.special.gammaln`

### CODE REVIEW

**Comment by StefanUlbrich** [comment]
Test passed successfully. @FelixWick I assume that means the workaround is functional? 
Shall we include [benchmark in our tests](https://pytest-benchmark.readthedocs.io/en/stable/) to catch performance regressions in the future?

**Comment by StefanUlbrich** [comment]
Included an error in the function and the other tests passed. Can we adjust testing for that?

**Comment by FelixWick** [comment]
It would be great to include performance tests, yes.

I would have expected the integration test for the width model to fail if you include an error in gammaln.
Also, it should at least give a different result if you require an integer input in your implementation.

**Comment by FelixWick** [comment]
That looks promising. Please go ahead with an implementation of your proposed function in the package. If you could add a performance check on the integration test of the width mode, along with some variations to make sure we get compatible results, that would be awesome.

**Comment by FelixWick** [comment]
Is this pull request ready for review and merge?

File: tests/test_gammaln.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+from scipy import special
+import numpy as np
+
+
+def test_gammaln():
+    data = np.linspace(1e-04, 2 - 1e-02, 100)
+    tol = (GAMMALN_CONSTANT + 1) * np.exp(-GAMMALN_CONSTANT)
+    for x in data:
+        assert abs(gammaln(x) - special.gammaln(x)) < tol, f"for x={x}"
+
```
Comment (FelixWick) [inline tests/test_gammaln.py:12]: What about the range between 2 and 10?

File: examples/gammaln_test.ipynb
Comment (FelixWick) [inline examples/gammaln_test.ipynb:1]: Can you plot for higher values, too?

**Comment by FelixWick** [comment]
I have checked the resulting values for c compared to master (using numba_scipy) and there are quite some deviations (mean of c in integration test increases from 0.365 to 0.391). Can we make numba_scipy an optional dependency for the ones who need a (potentially) more accurate width model and otherwise use the (potentially) less accurate implementation in this pull request?

**Review by FelixWick — CHANGES_REQUESTED** [review]
- please check comparison for higher values
- please make use of numba_scipy optional

**Comment by FelixWick** [comment]
@StefanUlbrich any progress on this?

**Comment by StefanUlbrich** [comment]
Apologies, still on it and will continue over the weekend—I also found a better pure python version that I want to try out.

**Comment by StefanUlbrich** [comment]
I have not been able to figure out how to specify `scipy-numba` as an optional / package extra. 
Poetry has problems ignoring the dependency for python 3.10+ ... Yet the code checks whether it is installed, so we could have the users take care of this by manually installing it.

The fast and accurate implementation I mentioned is [here](https://stackoverflow.com/a/54851190) and looks simple enough to implement. @FelixWick what do you recommend? The implementation looks accurate enough and straight forward

**Comment by FelixWick** [comment]
So, you're saying you want to replace your current algorithm with that one? I'm ok with that, of course, as long as it's faster and more accurate.

**Comment by StefanUlbrich** [comment]
> So, you're saying you want to replace your current algorithm with that one? I'm ok with that, of course, as long as it's faster and more accurate.

How did you test both? Do you have code suitable for a unit test? It will not be faster as the one in scipy—that is highly optimized after all—but it is better than sticking with the old numba-scipy.

**Comment by FelixWick** [comment]
Ok, maybe this is a misunderstanding :). I thought you're suggesting yet another implementation (from your stackoverflow link). So, that one is the same you already implemented before?

**Comment by StefanUlbrich** [comment]
> Ok, maybe this is a misunderstanding :). I thought you're suggesting yet another implementation (from your stackoverflow link). So, that one is the same you already implemented before?

It is a different one (it's iterative) but seems to be on par in terms of accuracy and close to performance with scipy from what I read

**Comment by FelixWick** [comment]
Ok, that's what I understood, good. No, I don't have a dedicated test. I think your checks in the notebook are just fine. Additionally, I used an integration test to check the consequences for the negative binomial overdispersion parameter in the end. These were quite large for your current implementation. Would be interesting to compare with the new one.

**Comment by FelixWick** [comment]
I think I found a much simpler solution :), see https://github.com/Blue-Yonder-OSS/cyclic-boosting/pull/43.

**Pull request closed without merging**